deployments/migrations/0073_auto_20230301_1606.py

```diff
@@ -0,0 +1,77 @@
+# Generated by Django 3.2.18 on 2023-03-01 16:06
+
+from django.db import migrations, models
+import django.db.models.deletion
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('deployments', '0072_sectortag'),
+    ]
+
+    operations = [
+        migrations.RunSQL(
+            sql=[
+                 ("INSERT INTO deployments_sector   (id, \"order\", slug, title) values (0 ,0 ,'WASH', 'WASH')"),
+                 ("INSERT INTO deployments_sector   (id, \"order\", slug, title) values (1 ,1 ,'PGI', 'PGI')"),
+                 ("INSERT INTO deployments_sector   (id, \"order\", slug, title) values (2 ,2 ,'CEA' , 'CEA')"),
+                 ("INSERT INTO deployments_sector   (id, \"order\", slug, title) values (3 ,3 ,'MIGRATION', 'Migration')"),
```

Comment:
We are doing some complex inserts. E.g. from array like {2,5,8} we create unnested ones and inserts them to another table.
Basically: we save the existing (and Enum-based) data.

Reply:
I am still confused. Can't we use fixtures for this? It will be easier in the future as well if we plan to update them.
We can create a script like init.py which will create/update data for these kinds of dynamic-static tables.

Reply:
The sequence of inserts can be done via fixtures. If you make it, I won't be against. But that complex task, reading from a {x, y, .. } array of sectortags, which should be unnested and loaded into another table, is hardly possible with nice little django commands. So some raw sql is unavoidable. And if there are some, then it's usable to keep all of them together.
Next time, when I will have the task to load a table and nothing else, I will use fixtures.

Reply:
If it works then it's okay for now.

Just dropping it here to show how we could also do this in a Django way.
NOTE: Reverse data migration is not working
```python
# Generated by Django 3.2.18 on 2023-03-01 16:06

from django.db import migrations, models
from django.contrib.postgres.aggregates import ArrayAgg
import django.db.models.deletion


class ExistingEnumSnapshot:
    class Sectors(models.IntegerChoices):
        WASH = 0, 'WASH'
        PGI = 1, 'PGI'
        CEA = 2, 'CEA'
        MIGRATION = 3, 'Migration'
        HEALTH = 4, 'Health'
        DRR = 5, 'DRR'
        SHELTER = 6, 'Shelter'
        NS_STRENGTHENING = 7, 'NS Strengthening'
        EDUCATION = 8, 'Education'
        LIVELIHOODS_AND_BASIC_NEEDS = 9, 'Livelihoods and basic needs'

    class SectorTags(models.IntegerChoices):
        WASH = 0, 'WASH'
        PGI = 1, 'PGI'
        CEA = 2, 'CEA'
        MIGRATION = 3, 'Migration'
        DRR = 5, 'DRR'
        SHELTER = 6, 'Shelter'
        NS_STRENGTHENING = 7, 'NS Strengthening'
        EDUCATION = 8, 'Education'
        LIVELIHOODS_AND_BASIC_NEEDS = 9, 'Livelihoods and basic needs'
        RECOVERY = 10, 'Recovery'
        INTERNAL_DISPLACEMENT = 11, 'Internal displacement'
        HEALTH_PUBLIC = 4, 'Health (public)'
        HEALTH_CLINICAL = 12, 'Health (clinical)'
        COVID_19 = 13, 'COVID-19'


def migrate_data(apps, schema_editor):
    Project = apps.get_model('deployments', 'Project')
    Sector = apps.get_model('deployments', 'Sector')
    SectorTag = apps.get_model('deployments', 'SectorTag')

    # Create Sector and SectorTags using pre-existing static choices
    new_sectors = []
    for value, label in ExistingEnumSnapshot.Sectors.choices:
        new_sectors.append(
            Sector(
                id=value,
                title=label,
            )
        )
    Sector.objects.bulk_create(new_sectors, ignore_conflicts=True)

    new_sector_tags = []
    for value, label in ExistingEnumSnapshot.SectorTags.choices:
        new_sector_tags.append(
            SectorTag(
                id=value,
                title=label,
            )
        )
    SectorTag.objects.bulk_create(new_sector_tags, ignore_conflicts=True)

    # Migrate project data
    Project.objects\
        .filter(primary_sector_legacy__isnull=False)\
        .update(primary_sector_id=models.F('primary_sector_legacy'))

    Project.secondary_sectors.through.objects.bulk_create([
        Project.secondary_sectors.through(
            project_id=project_id,
            sectortag_id=sector_tag_value,
        )
        for project_id, sector_tag_value in Project.objects
        .annotate(secondary_sector=models.functions.Cast(
            models.Func(
                models.F('secondary_sectors_legacy'),
                function='UNNEST'
            ),
            output_field=models.IntegerField(),
        )).values_list('id', 'secondary_sector')
    ])


def reverse_migrate_data(apps, schema_editor):
    # FIXME: THIS DOESTN"T WORK
    Project = apps.get_model('deployments', 'Project')

    # Migrate project data
    Project.objects\
        .filter(primary_sector_id__isnull=False)\
        .update(primary_sector_legacy=models.F('primary_sector_id'))

    Project.objects.bulk_update([
        Project(
            id=project_id,
            secondary_sectors_legacy=secondary_tags_value,
        )
        for project_id, secondary_tags_value in Project.secondary_sectors
        .through.objects.order_by().values('project_id').annotate(
            secondary_sectors_value=ArrayAgg('sectortag_id'),
        ).values_list('project_id', 'secondary_sectors_value')
    ], fields=('secondary_sectors_legacy',))

    # Delete M2M Rows
    Project.objects.filter(primary_sector_id=None)
    Project.secondary_sectors.through.objects.all().delete()


class Migration(migrations.Migration):

    dependencies = [
        ('deployments', '0072_sectortag'),
    ]

    operations = [
        # Rename existing fields
        migrations.RenameField(
            model_name='project',
            old_name='primary_sector',
            new_name='primary_sector_legacy',
        ),
        migrations.RenameField(
            model_name='project',
            old_name='secondary_sectors',
            new_name='secondary_sectors_legacy',
        ),
        # Add new fields
        migrations.AddField(
            model_name='project',
            name='primary_sector',
            field=models.ForeignKey(
                on_delete=django.db.models.deletion.CASCADE,
                to='deployments.sector',
                verbose_name='sector',
                null=True,
            ),
        ),
        migrations.AddField(
            model_name='project',
            name='secondary_sectors',
            field=models.ManyToManyField(
                blank=True,
                related_name='tags',
                to='deployments.SectorTag',
            ),
        ),
        migrations.AlterModelOptions(
            name='sector',
            options={
                'verbose_name': 'Project Sector',
                'verbose_name_plural': 'Project Sectors'
            },
        ),
        migrations.AlterModelOptions(
            name='sectortag',
            options={
                'verbose_name': 'Project Sector Tag',
                'verbose_name_plural': 'Project Sector Tags'
            },
        ),
        migrations.RunPython(
            migrate_data,
            # reverse_code=reverse_migrate_data,
            reverse_code=migrations.RunPython.noop,
        ),
        # Remove existing fields
        migrations.RemoveField(
            model_name='project',
            name='primary_sector_legacy',
        ),
        migrations.RemoveField(
            model_name='project',
            name='secondary_sectors_legacy',
        ),
        # Alter null=False
        migrations.AlterField(
            model_name='project',
            name='primary_sector',
            field=models.ForeignKey(
                on_delete=django.db.models.deletion.CASCADE,
                to='deployments.sector',
                verbose_name='sector',
                null=False,
            ),
        ),
    ]
```

Reply:
Great job! And a brilliant excercise... Thanks Navin!